src/routers/alpha-router/functions/get-candidate-pools.ts

```diff
@@ -941,5 +960,232 @@ export async function getV2CandidatePools({
     },
   };
 
-  return { poolAccessor, candidatePools: poolsBySelection };
+  return { poolAccessor, candidatePools: poolsBySelection, subgraphPools };
+}
+
+export async function getMixedRouteCandidatePools({
+  tokenIn,
+  tokenOut,
+  routeType,
+  routingConfig,
+  V3subgraphProvider,
+  V2subgraphProvider,
+  tokenProvider,
+  V3poolProvider,
+  V2poolProvider,
+  blockedTokenListProvider,
+  chainId,
+}: MixedRouteGetCandidatePoolsParams): Promise<{
+  V2poolAccessor: V2PoolAccessor;
+  V3poolAccessor: V3PoolAccessor;
+  candidatePools: CandidatePoolsBySelectionCriteria;
+  subgraphPools: (V2SubgraphPool | V3SubgraphPool)[];
+}> {
+  const { blockNumber } = routingConfig;
+  const { subgraphPools: V3subgraphPools, candidatePools: V3candidatePools } =
+    await getV3CandidatePools({
+      tokenIn,
+      tokenOut,
+      tokenProvider,
+      blockedTokenListProvider,
+      poolProvider: V3poolProvider,
+      routeType,
+      subgraphProvider: V3subgraphProvider,
+      routingConfig,
+      chainId,
+    });
+  const { subgraphPools: V2subgraphPools, candidatePools: V2candidatePools } =
+    await getV2CandidatePools({
+      tokenIn,
+      tokenOut,
+      tokenProvider,
+      blockedTokenListProvider,
+      poolProvider: V2poolProvider,
+      routeType,
+      subgraphProvider: V2subgraphProvider,
+      routingConfig,
+      chainId,
+    });
+
+  /**
+   * Main heuristic for pruning mixedRoutes:
+   * - we pick V2 pools with higher liq than respective V3 pools, or if the v3 pool doesn't exist
+   *
+   * This way we can reduce calls to our provider since it's possible to generate a lot of mixed routes
+   */
+  /// We only really care about pools involving the tokenIn or tokenOut explictly,
+  /// since there's no way a long tail token in V2 would be routed through as an intermediary
+  const V2topByTVLPools = [
+    ...V2candidatePools.selections.topByTVLUsingTokenIn,
+    ...V2candidatePools.selections.topByBaseWithTokenIn,
+    /// tokenOut:
+    ...V2candidatePools.selections.topByTVLUsingTokenOut,
+    ...V2candidatePools.selections.topByBaseWithTokenOut,
+    /// Direct swap:
+    ...V2candidatePools.selections.topByDirectSwapPool,
+  ];
+
+  const V2topByTVLSortedPools = _(V2subgraphPools)
+    .filter((pool) => V2topByTVLPools.map((p) => p.id).includes(pool.id))
```

Comment:
mapping V2TopByTVLPools is repeated work here. Should precompute it. Also should use a Set<> instead of doing array includes

Reply:
Wdm by precompting it?

Reply:
```  
const V2topByTVLSortedPools = _(V2subgraphPools)
    .filter((pool) => V2topByTVLPools.map((p) => p.id).includes(pool.id))
    .sortBy((pool) => -pool.reserveUSD)
    .value();
```
We want all of the `V2SubgraphPools` has type `SubgraphPool[]`, and we want to filter down all of them based on the `PoolId` we selected in building `V2topByTVLPools`. Then, we want to sort them by reserveUSD.

I can change it to this:
```
  const V2topByTVLPools = new Set([
    ...V2candidatePools.selections.topByTVLUsingTokenIn,
    ...V2candidatePools.selections.topByBaseWithTokenIn,
    /// tokenOut:
    ...V2candidatePools.selections.topByTVLUsingTokenOut,
    ...V2candidatePools.selections.topByBaseWithTokenOut,
    /// Direct swap:
    ...V2candidatePools.selections.topByDirectSwapPool,
  ]);

  const V2topByTVLSortedPools = _(V2subgraphPools)
    .filter((pool) =>
      V2topByTVLPools.has({
        id: pool.id,
      })
    )
    .sortBy((pool) => -pool.reserveUSD)
    .value();
```

Reply:
hmm that actually won't work because Set.has doesn't work like that for objects

Reply:
```  
const V2topByTVLPoolIds = new Set(
    [
      ...V2candidatePools.selections.topByTVLUsingTokenIn,
      ...V2candidatePools.selections.topByBaseWithTokenIn,
      /// tokenOut:
      ...V2candidatePools.selections.topByTVLUsingTokenOut,
      ...V2candidatePools.selections.topByBaseWithTokenOut,
      /// Direct swap:
      ...V2candidatePools.selections.topByDirectSwapPool,
    ].map((poolId) => poolId.id)
  );

  const V2topByTVLSortedPools = _(V2subgraphPools)
    .filter((pool) => V2topByTVLPoolIds.has(pool.id))
    .sortBy((pool) => -pool.reserveUSD)
    .value();
```